### DISCUSSION TITLE
What is the algorithms of EntityRuler or PhraseMatcher ?

### DISCUSSION THREAD

**Discussion by zhangweida2080** [explosion/spaCy#10956] [Help: Coding & Implementations]
I am using the **PhraseMatcher** and **EntityRuler**, but what is the matching algorithm under these two modules?  
It likes a black-box.
However, for research, I need to modify them or to understand the  implementation algorithm for answering questions. 
Is it using **forward Maximum Matching**  or not ? 
Thanks a lot.

**Comment by polm** [comment]
spaCy is open source, so actually it's not like a black box, and you can look at the source for all the parts. For example the PhraseMatcher source is in [this directory](https://github.com/explosion/spaCy/tree/master/spacy/matcher). 

Forward Maximum Matching is a method for text segmentation, but the PhraseMatcher isn't performing text segmentation, it's just labelling text.
